### PR TITLE
Add codesigntool to handle signing identity

### PR DESCRIPTION
**Pull request by ffreling**
This is a draft to engage conversation on the issue #172 

In order to support profiles with multiple identities, I replaced the shell command managing signing identites with a python tool that wraps `codesign`.

Naming can definitely be improved.

I am looking forward to your feedback.

### CODE REVIEW

**Comment by ffreling** [comment]
`I signed it!`

**Comment by thomasvl** [comment]
So the intent is rather then to pass the identity to use, to instead of it auto figured out during the signing?

If someone were using a remote cache, wouldn't this get into trouble since without the extra args, signing by different folks would be different, but the args going into the cache keys will be the same.

**Comment by steeve** [comment]
Yes this is one the issues raised in #172: signing without specifying an identity is not reproducible and thus should not be cached.
Also, the `process_and_sign` rule should perhaps be split in several actions.

**Comment by steeve** [comment]
Or perhaps the identity should be output in a specific file, which could then be read by the code signing tool. This way only the commands using the identity file would become naturally invalidated, should a remote cache be used with multi-identity mobileprovision.

**Comment by thomasvl** [comment]
Looks like some of the test failures are real, the rules had some support for mocking signin for testing, so you'll likely need to ensure that still working.  Also I think it is the macOS command line rule that is an odd one, and doesn't use the normal factory for making it's attributes, so you likely will have to add the new tool directly to it's attributes.

As far as the caching issue, I don't think there is a way to mark something as "don't cache" from bazels pov.  So splitting out an action to extract the identity into a file and then passing that into the next action to give bazel some exposure would likely be a good thing.

That split would also help with something we've been debating; when collecting the swift support libraries, we can tell it to resign them in advance for us, but we need the identity to pass, so having an action to do the extraction could be leveraged for that also.

**Comment by ffreling** [comment]
@thomasvl sorry for the delay, I was a bit busy lately.

As @steeve said, our objective is two-fold:

1) filter the possible signing identities to only select one we can sign with

The previous behavior was to select the first identity from the provisioning
profile. Now we take the first one that has a matching certificate.

2) replace the signing logic written in shell with a python tool to make it more
readable

Indeed with this approach, we broke the caching mecanism because with the same
input, we might use a separate signing identity.

We could either split the bundling process in 3:
1. unpack (+ optional archive post processing)
2. sign (with `no-cache`)
3. compress and pack

The signing stage would be explicitely tagged `no-cache`.

Or we could export the signing identity in its own file, invalidating the cache
for commands using the identity (in our case, the signing tool).

Is there a preferred approach?

**Comment by thomasvl** [comment]
No worries, the busy thing hit on this end also.  😄 

Forgot to mention before, the fact that bundling and signing is monolithic is on our radar, it is becoming more of an issue for folks doing iterative development on a device.  So we're also starting to look at splitting this up to allow frameworks to be caches after they are signed, and reducing what has to be resigned on each _edit/build/debug_ cycle.

**Comment by steeve** [comment]
@thomasvl sorry for the long wait, but would it be a satisfying solution for you to write off the package + codesign task as `no-cache` ?

**Comment by steeve** [comment]
Namely the `ProcessAndSign` action

**Comment by steeve** [comment]
I signed it !

**Comment by ffreling** [comment]
I confirm that I am working with @steeve on this issue. I am ok with his changes.

**Comment by steeve** [comment]
Pushed a branch to fix a typo that would break signing for the simulator

**Comment by thomasvl** [comment]
From looking at the `178.patch` all the `From:` entries seem to have signed the CLA, so not sure why the bot is grumpy.

**Comment by thomasvl** [comment]
Looks like some of the test failures may still be real?

**Comment by steeve** [comment]
Looking into it !

**Comment by steeve** [comment]
I don't get why it's failing, as the the tool is declared in the `_bundling_tool_attributes`

**Comment by steeve** [comment]
Looks like I can reproduce the bug locally, will fix!

**Comment by thomasvl** [comment]
Looks like it is `macos_command_line_application` that doesn't use the factory, but it does use `bundler.run()`.  So you'll need to ensure the new attribute makes it in there.

**Comment by steeve** [comment]
Alright, found the issue, actually I improved the PR a bit. Tests looking good so far.

**Comment by steeve** [comment]
I've added `_codesigntool` to `_code_signing_attributes`. All the tests are passing except for this one:
```
//test:ios_test_runner_unit_test                                         FAILED in 639.5s
  /private/var/tmp/_bazel_steeve/bf36c2a2daf0f473d8970113dcf65891/execroot/build_bazel_rules_apple/bazel-out/darwin-fastbuild/testlogs/test/ios_test_runner_unit_test/test.log```


The log says:
```
Target //ios:PassingUnitTestWithHost up-to-date:
  test-bin/ios/PassingUnitTestWithHost.zip
  test-bin/ios/PassingUnitTestWithHost
INFO: Elapsed time: 177.774s, Critical Path: 175.93s
INFO: 9 processes: 2 local, 7 processwrapper-sandbox.
INFO: Build completed, 1 test FAILED, 18 total actions
//ios:PassingUnitTestWithHost                                            FAILED in 169.2s
  /private/var/tmp/_bazel_steeve/bf36c2a2daf0f473d8970113dcf65891/sandbox/darwin-sandbox/28/execroot/build_bazel_rules_apple/_tmp/b57706bd1cb92d6fe477b01655fae99d/_bazel_steeve/22fb67d085c87e18a9d8f32c81210ced/execroot/build_bazel_rules_apple_integration_tests/bazel-out/darwin-fastbuild/testlogs/ios/PassingUnitTestWithHost/test.log

Executed 1 out of 1 test: 1 fails locally.
INFO: Build completed, 1 test FAILED, 18 total actions```

However I can't check the sublog.

File: apple/bundling/rule_factory.bzl [hunk 1]
```diff
@@ -322,6 +322,11 @@ def _code_signing_attributes(code_signing):
       "_skip_simulator_signing_allowed": attr.bool(
           default=code_signing.skip_simulator_signing_allowed
       ),
+      "_codesigningtool": attr.label(
```
Comment (steeve) [inline apple/bundling/rule_factory.bzl]: Relevant part here @thomasvl

File: apple/bundling/macos_rules.bzl [hunk 2]
```diff
@@ -289,7 +289,7 @@ def _macos_command_line_application_impl(ctx):
   signing_commands = codesigning_support.signing_command_lines(
       ctx, [path_to_sign], None)
 
-  inputs = [ctx.file.binary]
+  inputs = [ctx.file.binary, ctx.executable._codesigningtool]
```
Comment (steeve) [inline apple/bundling/macos_rules.bzl]: Relevant part here @thomasvl

**Comment by steeve** [comment]
Ah crap wait, made a typo

**Comment by steeve** [comment]
@thomasvl you need to reset the CLA again..

**Comment by steeve** [comment]
I'm going to rebase this, but should the tests pass, are you satisfied with the solution so far @thomasvl ?

**Comment by steeve** [comment]
I'm also going to remove @ffreling commits (he works with me) so that the CLA doesn't resets everytime. Are you fine @ffreling ?

**Comment by ffreling** [comment]
@steeve I'm fine with that.

**Comment by thomasvl** [comment]
Hanging this off to @sergiocampama since he's been working on the bundler changes lately and starting to think about signing improvements.

**Comment by steeve** [comment]
Thanks guys!
We can open a new PR if the CLA resetting is a pain for everyone.

**Comment by steeve** [comment]
hey @sergiocampama, is there anything that's missing or blocking on your side for this to move forward ?

**Comment by ffreling** [comment]
Closing the issue, I'll let @steeve deal with it in his own PR.

**Pull request closed without merging**